### ISSUE TITLE
The output format of mkldnn conv is wrong when data_format is NHWC

### ISSUE THREAD

**Issue by baoachun** [PaddlePaddle/Paddle#38126]
![图片](https://user-images.githubusercontent.com/22114318/145959037-c35b64e0-6f9e-47da-849b-fafd4e854d47.png)

The problem can be reproduced according to this pr：https://github.com/PaddlePaddle/Paddle/pull/38107

Runing the following command to reproduce.
```
git clone the code and cmake with -DWITH_TESTING=ON
then running the following command

ctest -R test_mkldnn_conv_gelu_fuse_pass -V
```

**Comment by lidanqing-vv** [comment]
@jczaja Please comment on this issue. Otherwise that disabling PR will be merged.

**Comment by jczaja** [comment]
@baoachun I implemented NHWC support for oneDNN kernels in PaddlePaddle so I can offer some insight.
If PaddlePaddle  input is of layout NCHW then shape is [N,C,H,W]. If PaddlePaddle input is of layout NHWC then shape is [N,H,W,C]. For oneDNN kernels input regardless if it is NCHW or NHWC then shape is always [N,C,H,W]. So internally for NHWC models when there is a CPU op followed by oneDNN op then there is shape rotation from [N,H,W,C] to [N,C,H,W].  And  when there is oneDNN op followed by CPU op then there is shape roatation from  [N,C, H, W] to [N,H,W,C]. The thing is that this assumes that after oneDNN op there is some CPU op. When We implemented NHWC support then there was fetch op always the last one so shape rotation happened there. But I noticed that later on PaddlePaddle by default is skipping fetch op and user get access to data that has still oneDNN shape. So proper fix would be to add this shape rotation  in a mechanics where you get access to oneDNN tensor. Here is line calling shape rotation: 
https://github.com/PaddlePaddle/Paddle/blob/21b7ed3e8d91ffb2fab3e28367c6eacfa292855b/paddle/fluid/framework/data_layout_transform.cc#L189-L191

Some more information (design doc on NHWC in oneDNN kernels):
https://github.com/PaddlePaddle/docs/blob/develop/docs/design/mkldnn/nhwc/nhwc.md

**Comment by baoachun** [comment]
Hi @jczaja , at present, a user encounters the inference error problem, which is caused because conv2d does not support NHWC format.
![图片](https://user-images.githubusercontent.com/22114318/147805937-ab300300-a121-47a2-9055-ae8c7cb7e2d0.png)
<img width="687" alt="图片" src="https://user-images.githubusercontent.com/22114318/147806014-e89b587c-b149-4626-83e8-5ff91b7b2420.png">

[am_veo.tar.gz](https://github.com/PaddlePaddle/Paddle/files/7795413/am_veo.tar.gz)

**Comment by jczaja** [comment]
@baoachun Thanks for more information. I'm planning to work on this issue right after I finish current issue so I will start work on NHWC next week.

**Comment by jczaja** [comment]
@baoachun Just to let You know that I started to investigate this issue. Reproduction was achieved. Problem seems to be that oneDNN elementwise_add (X,Y)  kernel with Y being broadcasted , does not work properly for NHWC situation. I will write some more when I get more details.

**Comment by jczaja** [comment]
@baoachun Situation is as follows. There are two problems encountered:
1) broadcasting  in oneDNN's elementwise kernels for NHWC is not working properly. Candidate fix is here : #38126 .
2) batch norm of NHWC in oneDNN's kernel does not work properly for 3D data. Status: under investigaton

**Comment by jczaja** [comment]
@baoachun I just rebased https://github.com/PaddlePaddle/Paddle/pull/39097 and it does make this issue to go away. So please test it if this PR #39097 works for you.

**Comment by baoachun** [comment]
Hi @jczaja , most of the UTs have passed, except for `test_mkldnn_conv_elementwise_add_fuse_pass`, the error message is as follows. I noticed that the default input in the conv operator is NCHW format. Is this judgment reasonable? https://github.com/PaddlePaddle/Paddle/blob/0d46a1085f58d20e8b9d3693172d5739e73cc08d/paddle/fluid/operators/conv_op.cc#L60
![图片](https://user-images.githubusercontent.com/22114318/154022848-7950124e-455a-4c93-b7a4-6bac378ebab4.png)

**Comment by jczaja** [comment]
@baoachun Ok, I understand that in https://github.com/PaddlePaddle/Paddle/pull/38107 there is UT added with ignored NHWC and I should enable and fix that one. I will start working on this.

**Comment by jczaja** [comment]
@baoachun Not sure if I'm solving proper issue, but I enabled NHWC part of UT in test checking fusing pass of conv and gelu. PR is here : https://github.com/PaddlePaddle/Paddle/pull/39591 . Check if this is what you have expected to fix?

**Comment by baoachun** [comment]
> Hi @jczaja , most of the UTs have passed, except for `test_mkldnn_conv_elementwise_add_fuse_pass`, the error message is as follows. I noticed that the default input in the conv operator is NCHW format. Is this judgment reasonable?
> 
> https://github.com/PaddlePaddle/Paddle/blob/0d46a1085f58d20e8b9d3693172d5739e73cc08d/paddle/fluid/operators/conv_op.cc#L60
> 
> ![图片](https://user-images.githubusercontent.com/22114318/154022848-7950124e-455a-4c93-b7a4-6bac378ebab4.png)

Hi @jczaja, I have removed most of the skip settings for NHWC, pr is here: https://github.com/PaddlePaddle/Paddle/pull/39551, but the `test_mkldnn_conv_elementwise_add_fuse_pass` UT failed, ans I got the above error message, I'm not sure this judgment is reasonable at this time.

**Comment by jczaja** [comment]
@baoachun Ok, so to reproduce this problem I should use enable NHWC in test_mkldnn_conv_elementwise_add_fuse_pass ?

**Comment by baoachun** [comment]
Hi @jczaja, you can use this pr https://github.com/PaddlePaddle/Paddle/pull/39654 to reproduce.

**Comment by jczaja** [comment]
The issues reported here are fixed by #40049 . Please retest and let us now if all works for you

**Comment by lidanqing-vv** [comment]
@baoachun said PaddleHub reported failures after the PR merged. Hence please keep this issue open. More info will follow

**Comment by lidanqing-vv** [comment]
To summarize:
1. The model error has been fixed by #40049, However, a newer Baidu commit broke the model again https://github.com/PaddlePaddle/Paddle/issues/40540
2. We expect the commit author could fix the issue, if you need any support or explanation of #40049, @jczaja will answer in time.

**Comment by yaomichael** [comment]
notes from 5/20 meeting:
The regression caused by PaddleHub team should have been fixed already. @jiangjiajun will check internally and close this ticket.
